src/Functions/seriesDecomposeSTL.cpp

```diff
@@ -0,0 +1,264 @@
+                || executeNumber<Int32>(src_data, period, prev_src_offset, curr_offset, seasonal, trend, residue)
+                || executeNumber<Int64>(src_data, period, prev_src_offset, curr_offset, seasonal, trend, residue)
+                || executeNumber<Float32>(src_data, period, prev_src_offset, curr_offset, seasonal, trend, residue)
+                || executeNumber<Float64>(src_data, period, prev_src_offset, curr_offset, seasonal, trend, residue))
+            {
+                res_data.insert(res_data.end(), seasonal.begin(), seasonal.end());
+                col_offsets_data.push_back(res_data.size());
+
+                res_data.insert(res_data.end(), trend.begin(), trend.end());
+                col_offsets_data.push_back(res_data.size());
+
+                res_data.insert(res_data.end(), residue.begin(), residue.end());
+                col_offsets_data.push_back(res_data.size());
+
+                root_offsets_data.push_back(col_offsets->size());
+
+                prev_src_offset = curr_offset;
+            }
+            else
+                throw Exception(
+                    ErrorCodes::ILLEGAL_COLUMN,
+                    "Illegal column {} of first argument of function {}",
+                    arguments[0].column->getName(),
+                    getName());
+        }
+        ColumnArray::MutablePtr nested_array_col = ColumnArray::create(std::move(ret), std::move(col_offsets));
+        return ColumnArray::create(std::move(nested_array_col), std::move(root_offsets));
+    }
+
+    template <typename T>
+    bool executeNumber(
+        const IColumn & src_data,
+        Float64 period,
+        ColumnArray::Offset & start,
+        ColumnArray::Offset & end,
+        std::vector<Float32> & seasonal,
+        std::vector<Float32> & trend,
+        std::vector<Float32> & residue) const
+    {
+        const ColumnVector<T> * src_data_concrete = checkAndGetColumn<ColumnVector<T>>(&src_data);
+        if (!src_data_concrete)
+            return false;
+
+        const PaddedPODArray<T> & src_vec = src_data_concrete->getData();
+
+        chassert(start <= end);
+        size_t len = end - start;
+        if (len < 4)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "At least four data points are needed for function {}", getName());
+        else if (period > (len / 2.0))
+            throw Exception(
+                ErrorCodes::BAD_ARGUMENTS, "The series should have data of at least two period lengths for function {}", getName());
+
+        std::vector<float> src((src_vec.begin() + start), (src_vec.begin() + end));
+
+        try
+        {
+            auto res = stl::params().fit(src, static_cast<size_t>(std::round(period)));
+
+            if (res.seasonal.empty())
+                return false;
+
+            seasonal = res.seasonal;
+            trend = res.trend;
+            residue = res.remainder;
+            return true;
+        }
+        catch (...)
```

Comment:
Exception-to-error-code translation (or even swallowing exceptions) is a code smell. Let's remove the try/catch and have exceptions bubble up to the caller.